### PR TITLE
vdev_mirror: when resilvering, try reading first

### PR DESCRIPTION
**Pull request by nwf**
### Motivation and Context

From experience, resilvers can make it most of the way through and die for one reason or another.

### Description

Optimistically, then, fling reads at devices being resilvered and respond to checksum mismatches rather than always writing.

As this is a kind of nop-write, and out of an abundnace of caution, restrict this to checksum types strong enough to be considered good enough.  (This, notably, excludes fletcher4, which is always used for pool metadata.)

Signed-off-by: Nathaniel Wesley Filardo <nwfilardo@gmail.com>

### How Has This Been Tested?

Local use with `zhack scrub`.

### Types of changes
- [x] Performance enhancement (non-breaking change which improves efficiency)

### Checklist:
- [ ] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by jumbi77** [comment]
Just referencing an old Illumos / OpenZFS ticket [5532](https://www.illumos.org/issues/5532), which never get upstreamed as far as I can tell. (@amotin)

**Comment by nwf** [comment]
> Just referencing an old Illumos / OpenZFS ticket [5532](https://www.illumos.org/issues/5532), which covers similar code lines..but not really sure if its applicable...

Looks still applicable to me.  Adding a test for `mm->children > 1` to the conditional here will cause the `mm->children == 1` case to fall through to `vdev_mirror_child_select`, which will pick the one child (barring badness) and reach the later `zio` calls in `vdev_mirror_io_start`, which avoids `abd_alloc_sametype`, `bcopy`, and `abd_free`.  In fact it might be nice to reach that case if there's one child left standing after the test introduced by #11930, too.

File: module/zfs/vdev_mirror.c [hunk 1]
```diff
@@ -637,8 +638,15 @@ vdev_mirror_io_start(zio_t *zio)
...
 	if (zio->io_type == ZIO_TYPE_READ) {
-		if (zio->io_bp != NULL &&
-		    (zio->io_flags & ZIO_FLAG_SCRUB) && !mm->mm_resilvering) {
+		boolean_t have_checksum = zio->io_bp != NULL;
+		boolean_t is_scrub = have_checksum &&
+		    (zio->io_flags & ZIO_FLAG_SCRUB) &&
+		    !mm->mm_resilvering;
+		boolean_t resilver_read = have_checksum &&
+		    (zio->io_flags & (ZIO_FLAG_SCRUB | ZIO_FLAG_RESILVER)) &&
+		    (zio_checksum_table[BP_GET_CHECKSUM(zio->io_bp)].ci_flags &
```
Comment (ahrens) [inline module/zfs/vdev_mirror.c:647]: This would be a good place to add a comment explaining the reasoning behind this.

**Review by ahrens — COMMENTED** [review]
I think this is saying that when we resilver, we read from both sides (old and new), and then write to the new side.  Does this also short-circuit the write if the read from the new side checksums correctly?  If so, that means that a resilver would read the entire new disk and then also write it, making resilvers go about half as fast?  What's the benefit behind this?

**Comment by nwf** [comment]
For some drive technologies, most notably SMR, writing can be significantly slower than reading and this can significantly improve recovering from aborted resilvers or from the ill effects of pre-#11930 scans of mirrors with offline devices.

That said, it's probably a good idea to request this read-before-write strategy only on some scans and on some pools, so that we don't pay the read-then-write slowdown when we don't expect the new side to already be full of the correct data.  That might mean a new ZIO flag that's controlled by the scan logic.  I'll flesh the proposal out to include that if you think this change isn't a challenge to data integrity.

**Comment by jumbi77** [comment]
> > Just referencing an old Illumos / OpenZFS ticket [5532](https://www.illumos.org/issues/5532), which covers similar code lines..but not really sure if its applicable...
> 
> Looks still applicable to me. Adding a test for `mm->children > 1` to the conditional here will cause the `mm->children == 1` case to fall through to `vdev_mirror_child_select`, which will pick the one child (barring badness) and reach the later `zio` calls in `vdev_mirror_io_start`, which avoids `abd_alloc_sametype`, `bcopy`, and `abd_free`. In fact it might be nice to reach that case if there's one child left standing after the test introduced by #11930, too.

Will you then add `mm->mm_children > 1` ?

**Comment by jumbi77** [comment]
@nwf Can you may rebase this, add the `mm->mm_children > 1` check and resolve the feedback (add requested comment)? Much thanks in advance and looking forward to get this upstreamed.

**Comment by jumbi77** [comment]
> @nwf Can you may rebase this, resolve the feedback (add requested comment) and optionally add the `mm->mm_children > 1` check ? Much thanks in advance and looking forward to get this upstreamed.
> 
> @mmaybee Can you may take a look too? Thanks in advance for your review.

Politely ping @nwf. Do you may can rebase this? Much thanks in advance.

**Comment by amotin** [comment]
I think we should address resilver restarts issue via reducing the amount of block pointer we needlessly accumulate in RAM for sorting if the pool is not fragmented, so that we could create checkpoints more often. I don't think it is reasonable to expect during general case resilver (unlike scrub) for the target device to already have all the data.  I agree with @ahrens that this would take at least twice more time for new disks.

**Comment by jumbi77** [comment]
> I think we should address resilver restarts issue via reducing the amount of block pointer we needlessly accumulate in RAM for sorting if the pool is not fragmented, so that we could create checkpoints more often. I don't think it is reasonable to expect during general case resilver (unlike scrub) for the target device to already have all the data.  I agree with @ahrens that this would take at least twice more time for new disks.

> And is just the one liner ```mm->mm_children``` from your original illumos ticket applyable (any benefit without side effect?)

**Comment by amotin** [comment]
> And is just the one liner `mm->mm_children > 1` from your original illumos ticket applyable (any benefit without side effect?)

@jumbi77 If you are about avoiding extra memory copy, then it is already handled by https://github.com/openzfs/zfs/pull/13606 .

**Pull request closed without merging**